### ISSUE TITLE
A possible approach to the "lost timestamp" issue

### ISSUE THREAD

**Issue by jdtsmith** [casouri/vundo#66]
## Recap of the issue

A recap from #62: there is one type of information that is lost when vundo trims the "useless" tails of very long undo-buffer lists — _timestamps_.  Emacs inserts timestamps into the modification record when modifying a saved (aka unmodified) buffer.  These timestamp (TS) records are simple, and look like:

```elisp
 (t 25616 45086 191805 680000) ; the numbers are the same format as (current-time)
```

Importantly, timestamps are inserted whether the modification in question is an "ordinary" modification, or an "undo/redo" modification.  But vundo trims the latter to avoid the `buffer-undo-list` growing exponentially (which happens fast thanks to its ability to _fly_ around).  In doing this trimming, timestamps associated with undo/redo modifications are lost.

## Impact

These lost timestamps show up in two ways:

1. vundo and regular undo/redo differ in their ability to correctly mark a buffer as unmodified on returning to that state; vundo cannot always recognize when it has returned to an unmodified buffer state because of this lossage.  undo/redo never fails.
2. The new saved-state indication from #62 does not work when you undo to some older state, save it, then undo/redo away from it.  The TS that is introduced from this is currently culled by vundo, no trace remaining.

## Potential Solution

Here is a possible solution: for any new TS records found past the last ordinary modification, "teleport" it back to the equivalent node closest to the root with the same directionality.  Said more precisely: for `t` representing the newly discovered node with the timestamp, pair (`t-1`, `t`) is mapped to _equivalent pair_ (`p-1`, `p`), for the minimal `p`.  Then we _teleport_ the TS from `t`→`p`, overwriting any found there.

## Diagram

In picture form (borrowing from @casouri's [excellent blog post](https://archive.casouri.cc/note/2021/visual-undo-tree/)):

![image](https://user-images.githubusercontent.com/93749/225157072-981f1f40-2580-452f-b3a5-83ef6a13f969.png)

## Outcome

To evaluate if a given node is _saved_, we check if the next state of that node _or any equivalent node_ has a TS record.  For example, in the diagram, node 1 is known to be saved, because its equivalent node 5 sees a TS in the next state up (node 6). This will fix problem 2 above.  For problem 1, as long as vundo selects a path that arrives via the adjacent TS node, this will also be fixed.  So for example, in the above right-hand example, going from 0→1 is via 6→5, which works correctly.

## Cost

The cost is (sometimes) maintaining a bit more tail to track save-then-undo's. In the above example (pink node is current), normally you'd trim back to node 5.  For the left scenario, you do that.  For the right scenario, you trim only back to node 6.

**Comment by jdtsmith** [comment]
Valid point.  There is a config option to disable save marking in #62 if you want to avoid seeing that information for some reason.  It should not impact performance at all.  This particular change would need some testing, but I don't expect it would have much performance impact either, since it's a one-time cost not associated with moving around.  

Is there a test framework you or others have used to stress-test the undo system?

**Comment by jdtsmith** [comment]
Here is a small stress test I wrote of the undo and vundo system.  

```elisp
;; Stress tests of the undo/vundo system

(require 'lorem-ipsum)
(eval-when-compile
  (require 'cl-lib))
(require 'memory-report)
(defun vundo--stress-test (&optional nedits nundos max-chain-length)
  (interactive
   (list (read-number "Number of edits (sentences): " 3000)
	 (read-number "Approximate number of undo/redo chains: " 500)
	 (read-number "Maximum undo length: " 14)))  ;*very* sensitive to this number
  (with-current-buffer (get-buffer-create "*vundo-stress-test*")
    (setq buffer-undo-list nil
	  pending-undo-list nil)
    (erase-buffer)
    (cl-loop for n upto nedits
	     for r = (random nedits)
	     do
	     (lorem-ipsum-insert-sentences 1)
	     (insert "\n")
	     (undo-boundary)
	     (when (<= r nundos)
	       (condition-case nil
		   (progn (undo (1+  (random max-chain-length)))
			  (undo-boundary))
		 (error (undo-boundary)))))
    (display-buffer "*vundo-stress-test*")
    (message "Undo stress test complete:\n\t%d sentences inserted\n\t%d undo chains of max length %d"
	     nedits nundos max-chain-length)
    (message "\n\t buffer-undo-list = %d bytes" (memory-report-object-size buffer-undo-list))
    (message "Launching vundo: ")
    (vundo--debug)
    (benchmark-progn (vundo))))
```

Just `M-x vundo--stress-test` and answer prompts.  I tuned the defaults to produce about 10MB or so of data on the buffer-undo-list.  

It reveals how wickedly fast and efficient vundo is.  In playing with this, It's pretty hard to make vundo take longer than 0.5s to popup _on an 8yo laptop_.  It's much easier to hit undo's memory limits while constructing the buffer (you'll need to increase your undo limits).  Undo chain length is particularly sensitive.  After some experimentation, something like this might be a nice benchmark to include as various features are considered.   Produces some pretty wild looking trees too ;)...

![image](https://user-images.githubusercontent.com/93749/225513190-81168f22-7ee4-4e95-b26b-5dbb6c4d8ad8.png)

**Comment by jdtsmith** [comment]
While I am thinking of it: I want to be clear that I consider the "incorrect buffer unmodified state" (impact 1) to be the far more significant issue than "getting the last-saved highlighted node right at all times" (impact 2).  But the latter makes it quite clear that information loss is occurring.  

Here's a recipe to see this yourself (using #62):

- Create a file with a fairly long (linear) set of changes.
- Using vundo, undo once, and save that state.
- Re-enter vundo: the saved node is highlighted as latest.
- Navigate forward and backwards a few times past this node, finally returning to it.  Hit Return.
- The buffer is (often) marked as _modified_, despite having recently been saved.

If you had done the _precise same set_ of modifications and undos/redos with the built-in emacs commands, the saved buffer state would have been correctly marked as unmodified on returning to it.   This is the current price paid for "trimming the tail".  Also note that TS lossage does not _always_ happen, since vundo does save "some" of the undo/redo-only tail of buffer-undo-list, as needed.  So sometimes the undo/redo-inserted TS is retained, other times it is lost.  This of course makes the situation even more confusing.  I often experience a related dissonance using vundo — "Wait, I did save there, didn't I?".

**Comment by jdtsmith** [comment]
OK, I did some more exploration of the "timestamp lossage" issue.  I've understood a few new things:

1. Navigating back up the tree and saving _for the first time_ a particular state (as in the example above) does introduce new timestamps into `buffer-undo-list`.  But since they appear _only_ in the "tail of undos", they are quite likely to be trimmed away.  The recipe above demonstrates this clearly, and #73 includes some debug reporting of when this occurs.
2. For now our simple workaround is: if the current state is unmodified, mark it as latest saved.   This is quite fragile and will not persist as you navigate away from that state.
1. The same TS is injected multiple times into `buffer-undo-list` as you navigate, which makes sense, since an undo mod records that it changed the same saved state (again).  Many of these duplicate timestamps get trimmed away by vundo, but without apparent loss of information.
3. Some TS's are present in the `buffer-undo-list`, but are _not_ (currently) being displayed by vundo. This is because the TS is not attached to the "master" index, but to one of its (later) equivalents.  This would be the case if you returned to an old state, saved it, then introduced new changes.   The next index after an equivalent (not the master node) records that timestamp.

More info in #73.

**Comment by jdtsmith** [comment]
My basic understanding is that moving a big distance in the undo tree adds significant length to the buffer-undo-list, which may, prior to vundo lopping off the unnecessary "leafy growth" at the tips of the tree, lead to list trimming from near the _trunk_ by emacs. Then some branches of the tree (including the one which you may be on) can be severed and unreachable.  

You can likely increase the various undo limits to help avoid this.  But never totally solve. A smart pruning that respects the tree structure could help, but sounds pretty complicated. Another idea is to break large moves into a series of smaller ones, trimming the shrubbery at each intermediate step.  Of course if emacs had used a tree structure (instead of a list that loops back on itself) that would be simpler.

**Comment by casouri** [comment]
Yes, increasing `undo-limit` should help. I don't think moving by arrow keys would add more stuff to undo list tho.

**Comment by jdtsmith** [comment]
Current simplified thinking on this issue is in #81.

**Comment by mentalisttraceur** [comment]
@jdtsmith is this issue still an issue or did you fix it in #96 ?
